File: LinkedList/DoublyLinkedList.cpp
#include<iostream>
using namespace std;

class node{
    public:
        int data;
        node* prev;
        node* next;

        node() = default;

        node(int data){
            this->data = data;
            this->next = NULL;
            this->prev = NULL;
        }
};

class DoublyLinkedList{
    public:
        node *head, *tail;
        int size = 0;

        DoublyLinkedList(int data){
            head = NULL;
            tail = NULL;

            createNextNode(data);
        }

        void createNextNode(int data){
            node* newNode = new node(data);

            if(head == NULL){
                head = newNode;
                tail = newNode;
                newNode = NULL;
            }
            else{
                tail->next = newNode;
                newNode->prev = tail;

                tail = newNode;
            }

            size++;
        }

        void insertAtHead(int data){
            node* newNode = new node(data);
            newNode->next = head;
            
            head->prev = newNode;

            head = newNode;

            size++;
        }

        void insertAtTail(int data){
            createNextNode(data);
        }

        void insertAtPosition(int pos, int data){
            node* newNode = new node(data);
            node* temp = new node;
            temp = head;
            int cnt = 1;
            
            if(pos == 1){
                insertAtHead(data);
                return;
            }
            else if(pos == size || pos>size){
                insertAtTail(data);
                return;
            }

            while(cnt < pos-1){
                temp = temp->next;
                cnt++;
            }

            newNode->next = temp->next;
            temp->next->prev = newNode;
            newNode->prev = temp;
            temp->next = newNode;

            size++;
        }

        void deleteAtHead(){
            node* temp = new node;
            temp = head;

            head = head->next;
            head->prev = NULL;

            delete temp;

            // cout << head->prev->data;
            size--;
        }

        void deleteAtTail(){
            node* temp = new node;
            temp = tail;

            tail = tail->prev;
            tail->next = NULL;

            delete temp;

            // cout << tail->next;
            size--;
        }

        void deleteAtPosition(int pos){
            node* temp = new node;
            node* prev_node = new node;
            temp = head;
            int cnt = 1;

            if(pos == 1){
                deleteAtHead();
                return;
            }
            else if(pos == size || pos>size){
                deleteAtTail();
                return;
            }

            while(cnt <= pos-1){
                prev_node = temp;
                temp = temp->next;
                cnt++;
            }

            prev_node->next = temp->next;
            temp->next->prev = prev_node;

            delete temp;

            size--;
        }

        void print(){
            node* temp = new node;
            temp = head;

            while(temp != NULL){
                cout << temp->data << " - ";
                temp = temp->next;
            }
            cout << "NULL" << endl;
        }

        void printReverse(){
            node* temp = new node;
            temp = tail;

            while(temp != NULL){
                cout << temp->data << " - ";
                temp = temp->prev;
            }
            cout << "NULL" << endl;
        }

        void printSize(){
            cout << size << endl;
        }
};

int main(){

    DoublyLinkedList dl(10);
    dl.createNextNode(20);
    dl.createNextNode(30);

    cout << "Traverse from head : ";
    dl.print();

    cout << "Traverse from tail : ";
    dl.printReverse();

    cout << "Size : ";
    dl.printSize();

    dl.insertAtHead(1);
    dl.print();

    dl.insertAtTail(40);
    dl.print();

    dl.insertAtPosition(2, 50);
    dl.print();
    
    cout << "Size : ";
    dl.printSize();

    dl.deleteAtHead();
    dl.print();

    dl.deleteAtTail();
    dl.print();

    dl.deleteAtPosition(3);
    dl.print();
    dl.printReverse();
    dl.printSize();

    dl.deleteAtHead();
    dl.print();
    dl.insertAtPosition(2, 20);
    dl.createNextNode(40);
    dl.insertAtPosition(4,50);

    // dl.deleteAtHead();

    dl.print();

    return 0;
}

File: Program's_Contributed_By_Contributors/Python_Programs/doublyLinkedList.py
class Node:
    def __init__(self, value) -> None:
        self.value = value
        self.next = None
        self.prev = None


class DoublyLinkedList:
    def __init__(self, value) -> None:
        new_node = Node(value)
        self.head = new_node
        self.tail = new_node
        self.length = 1

    def pop_first(self):
        if self.length == 0:
            return None
        temp = self.head
        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            self.head = self.head.next
            self.head.prev = None
            temp.next = None
        self.length -= 1
        return temp

    def pop(self):
        if self.length == 0:
            return None
        temp = self.tail
        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            self.tail = self.tail.prev
            self.tail.next = None
            temp.prev = None
        self.length -= 1
        return temp

    def append(self, value):
        new_node = Node(value)
        if self.head is None:
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            new_node.prev = self.tail
            self.tail = new_node
        self.length += 1
        return True

    def print_list(self):
        temp = self.head
        print("[", end="")
        while temp:
            if temp.next != None:
                print(temp.value, end=", ")
            else:
                print(f"{temp.value}]")
            temp = temp.next

    def append(self, value):
        new_node = Node(value)
        if self.head is None:
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            new_node.prev = self.tail
            self.tail = new_node
        self.length += 1
        return True

    def prepend(self, value):
        new_node = Node(value)
        if self.length == 0:
            self.head = new_node
            self.tail = new_node
        else:
            new_node.next = self.head
            self.head.prev = new_node
            self.head = new_node
        self.length += 1
        return True

    def get(self, index):
        if index < 0 or index >= self.length:
            return None
        temp = self.head
        if index < self.length/2:
            for _ in range(index):
                temp = temp.next
        else:
            temp = self.tail
            for _ in range(self.length - 1, index, -1):
                temp = temp.prev
        return temp

    def set_value(self, index, value):
        temp = self.get(index)
        if temp:
            temp.value = value
            return True
        return False


my_doubly_linked_list = DoublyLinkedList(11)

print("")

print('before append:')
my_doubly_linked_list.print_list()
print('after append:')
my_doubly_linked_list.append(3)
my_doubly_linked_list.append(23)
my_doubly_linked_list.append(7)
my_doubly_linked_list.append(90)
my_doubly_linked_list.append(8)
my_doubly_linked_list.append(31)
my_doubly_linked_list.print_list()

print("")
print("")

print('before pop:')
my_doubly_linked_list.print_list()
print('after pop:')
my_doubly_linked_list.pop()
my_doubly_linked_list.print_list()

print("")
print("")

print('before prepend:')
my_doubly_linked_list.print_list()
my_doubly_linked_list.prepend(1)
print('after prepend:')
my_doubly_linked_list.print_list()

print("")
print("")

print('before pop_first:')
my_doubly_linked_list.print_list()
print('after pop_first:')
my_doubly_linked_list.pop_first()
my_doubly_linked_list.print_list()

print("")
print("")

print('get(2) -> should return 23:')
print(my_doubly_linked_list.get(2).value)

print("")
print("")

print('before set_value(2, 9):')
my_doubly_linked_list.print_list()
my_doubly_linked_list.set_value(2, 9)
print('after set_value(2, 9):')
my_doubly_linked_list.print_list()

print("")
print("")

try:
    print(f'Head: {my_doubly_linked_list.head.value}')
    print(f'Tail: {my_doubly_linked_list.tail.value}')
except:
    print(f'Head: {my_doubly_linked_list.head}')
    print(f'Tail: {my_doubly_linked_list.tail}')

print(f'Length: {my_doubly_linked_list.length}')
